core/scheduler/src/main/scala/org/apache/openwhisk/core/scheduler/container/ContainerManager.scala

```diff
@@ -462,32 +539,30 @@ object ContainerManager {
     list
   }
 
-  private def chooseInvokerFromCandidates(
-    candidates: List[InvokerHealth],
-    wholeInvokers: List[InvokerHealth],
-    pairs: List[ScheduledPair],
-    msg: ContainerCreationMessage)(implicit logging: Logging): (List[ScheduledPair], List[InvokerHealth]) = {
-    val idx = rng(mod = candidates.size)
-    val instance = candidates(idx)
-    // it must be compared to the instance unique id
-    val idxInWhole = wholeInvokers.indexOf(wholeInvokers.filter(p => p.id.instance == instance.id.instance).head)
-    val requiredMemory = msg.whiskActionMetaData.limits.memory.megabytes
-    val updated =
-      if (instance.id.userMemory.toMB - requiredMemory >= requiredMemory) { // Since ByteSize is negative, it converts to long type and compares.
-        wholeInvokers.updated(
-          idxInWhole,
-          instance.copy(id = instance.id.copy(userMemory = instance.id.userMemory - requiredMemory.MB)))
+  @tailrec
+  protected[container] def chooseInvokerFromCandidates(candidates: List[InvokerHealth], msg: ContainerCreationMessage)(
+    implicit logging: Logging): ScheduledPair = {
+    val requiredMemory = msg.whiskActionMetaData.limits.memory
+    if (candidates.isEmpty) {
+      ScheduledPair(msg, invokerId = None, Some(NoAvailableInvokersError), Some(s"No available invokers."))
+    } else if (candidates.forall(p => p.id.userMemory.toMB < requiredMemory.megabytes)) {
+      ScheduledPair(msg, invokerId = None, Some(NoAvailableInvokersError), Some(s"No available invokers."))
```

Comment:
Seems case1 and case2's reason is same `No available invokers`
But they has small difference, e.g. case2, No available invokers with lack memory.

Reply:
So do you suggest adding a new error?

Reply:
hm...no. i think it is ok.

Reply:
I changed the error to `ResourceNotEnoughError`